pkg/package-server/provider/inmem.go

```diff
@@ -0,0 +1,181 @@
+package provider
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/ghodss/yaml"
+	operatorsv1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/queueinformer"
+	packagev1alpha1 "github.com/operator-framework/operator-lifecycle-manager/pkg/package-server/apis/packagemanifest/v1alpha1"
+	"github.com/prometheus/common/log"
+)
+
+const ConfigMapPackageName = "packages"
+
+type packageKey struct {
+	catalogSourceName      string
+	catalogSourceNamespace string
+	packageName            string
+}
+
+type InMemoryProvider struct {
+	*queueinformer.Operator
+	// catalogSourceLister catalogv1alpha1.CatalogSourceLister
+
+	mu        sync.RWMutex
+	manifests map[packageKey]packagev1alpha1.PackageManifest
+}
+
+func NewInMemoryProvider(informers []cache.SharedIndexInformer, queueOperator *queueinformer.Operator) *InMemoryProvider {
+	// instantiate the in-mem provider
+	prov := &InMemoryProvider{
+		Operator:  queueOperator,
+		manifests: make(map[packageKey]packagev1alpha1.PackageManifest),
+	}
+
+	// register CatalogSource informers.
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "catalogsources")
+	queueInformers := queueinformer.New(
+		queue,
+		informers,
+		prov.syncCatalogSources,
+		nil,
+		"catsrc",
+	)
+	for _, informer := range queueInformers {
+		prov.RegisterQueueInformer(informer)
+	}
+
+	return prov
+}
+
+func loadPackageManifestsFromConfigMap(cm *corev1.ConfigMap) ([]packagev1alpha1.PackageManifest, error) {
+	var manifests []packagev1alpha1.PackageManifest
+
+	cmName := cm.GetName()
+	found := false
+	packageListYaml, ok := cm.Data[ConfigMapPackageName]
+	if ok {
+		log.Debug("Load ConfigMap      -- ConfigMap contains packages")
+		packageListJson, err := yaml.YAMLToJSON([]byte(packageListYaml))
+		if err != nil {
+			log.Debugf("Load ConfigMap     -- ERROR %s : error=%s", cmName, err)
+			return nil, fmt.Errorf("error loading package list yaml from ConfigMap %s: %s", cmName, err)
+		}
+
+		var parsedSpecs []packagev1alpha1.PackageManifestSpec
+		err = json.Unmarshal([]byte(packageListJson), &parsedSpecs)
+		if err != nil {
+			log.Debugf("Load ConfigMap     -- ERROR %s : error=%s", cmName, err)
+			return nil, fmt.Errorf("error parsing package list (json) from ConfigMap %s: %s", cmName, err)
+		}
+
+		for _, spec := range parsedSpecs {
+			found = true
+			manifest := packagev1alpha1.PackageManifest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      spec.PackageName,
+					Namespace: cm.GetNamespace(),
+				},
+				Spec: spec,
+			}
+			// TODO: add check for invalid package definitions
+			manifests = append(manifests, manifest)
+		}
+		log.Debugf("Load ConfigMap      -- Found packages: %v", manifests)
```

Comment:
since we're copying this over anyway, could we make this structured with `log.WithField` etc like we do for other parts of olm?

Reply:
Yup, sounds good.